File: crates/koloda-core/src/domain/progress.rs
//! Shared FSRS progress-field bounds used by cards and reviews.
//!
//! Callers pass entity-specific error codes so UI namespaces
//! (`validation.cards.progress.*` vs `validation.reviews.*`) stay distinct.

use crate::app::error::AppError;
use crate::domain::cards::CardState;

pub const DIFFICULTY_MIN: f64 = 0.0;
pub const DIFFICULTY_MAX: f64 = 10.0;

pub fn validate_state(state: i32, code: &str) -> Result<(), AppError> {
    if !CardState::is_valid(state) {
        return Err(AppError::new(
            code,
            Some(format!(
                "State must be between {} and {}, got {}",
                CardState::MIN,
                CardState::MAX,
                state
            )),
        ));
    }
    Ok(())
}

pub fn validate_stability(stability: f64, code: &str) -> Result<(), AppError> {
    if stability < 0.0 {
        return Err(AppError::new(
            code,
            Some(format!("Stability must be non-negative, got {}", stability)),
        ));
    }
    Ok(())
}

pub fn validate_difficulty(difficulty: f64, code: &str) -> Result<(), AppError> {
    if !(DIFFICULTY_MIN..=DIFFICULTY_MAX).contains(&difficulty) {
        return Err(AppError::new(
            code,
            Some(format!(
                "Difficulty must be between {} and {}, got {}",
                DIFFICULTY_MIN, DIFFICULTY_MAX, difficulty
            )),
        ));
    }
    Ok(())
}

pub fn validate_scheduled_days(days: i32, code: &str) -> Result<(), AppError> {
    if days < 0 {
        return Err(AppError::new(
            code,
            Some(format!("Scheduled days must be non-negative, got {}", days)),
        ));
    }
    Ok(())
}

pub fn validate_learning_steps(steps: i32, code: &str) -> Result<(), AppError> {
    if steps < 0 {
        return Err(AppError::new(
            code,
            Some(format!("Learning steps must be non-negative, got {}", steps)),
        ));
    }
    Ok(())
}

pub fn validate_reps(reps: i32, code: &str) -> Result<(), AppError> {
    if reps < 0 {
        return Err(AppError::new(
            code,
            Some(format!("Reps must be non-negative, got {}", reps)),
        ));
    }
    Ok(())
}

pub fn validate_lapses(lapses: i32, code: &str) -> Result<(), AppError> {
    if lapses < 0 {
        return Err(AppError::new(
            code,
            Some(format!("Lapses must be non-negative, got {}", lapses)),
        ));
    }
    Ok(())
}

File: libs/srs/src/lib/progress.ts
import type { z } from "zod";

/** Mirrors `koloda` `domain/progress.rs` bounds. */
export const DIFFICULTY_MIN = 0;
export const DIFFICULTY_MAX = 10;
export const CARD_STATE_MIN = 0;
export const CARD_STATE_MAX = 3;
export const RATING_MIN = 1;
export const RATING_MAX = 4;

export type ProgressFieldCodes = {
  state: string;
  stability: string;
  difficulty: string;
  scheduledDays: string;
  learningSteps: string;
  reps?: string;
  lapses?: string;
};

export type ProgressFieldValues = {
  state: number;
  stability: number;
  difficulty: number;
  scheduledDays: number;
  learningSteps: number;
  reps?: number;
  lapses?: number;
};

export const CARDS_PROGRESS_FIELD_CODES = {
  state: "validation.cards-progress.state",
  stability: "validation.cards-progress.stability",
  difficulty: "validation.cards-progress.difficulty",
  scheduledDays: "validation.cards-progress.scheduled-days",
  learningSteps: "validation.cards-progress.learning-steps",
  reps: "validation.cards-progress.reps",
  lapses: "validation.cards-progress.lapses",
} as const satisfies ProgressFieldCodes;

export const REVIEWS_PROGRESS_FIELD_CODES = {
  state: "validation.reviews.state",
  stability: "validation.reviews.stability",
  difficulty: "validation.reviews.difficulty",
  scheduledDays: "validation.reviews.scheduled-days",
  learningSteps: "validation.reviews.learning-steps",
} as const satisfies ProgressFieldCodes;

function addFieldIssue(ctx: z.RefinementCtx, path: (string | number)[], code: string) {
  ctx.addIssue({ code: "custom", message: code, path });
}

/** Shared FSRS progress-field bounds — callers supply entity-specific error codes. */
export function validateProgressFields(
  data: ProgressFieldValues,
  codes: ProgressFieldCodes,
  ctx: z.RefinementCtx,
  pathPrefix: (string | number)[] = [],
) {
  const fieldPath = (field: string) => [...pathPrefix, field];

  if (data.state < CARD_STATE_MIN || data.state > CARD_STATE_MAX) {
    addFieldIssue(ctx, fieldPath("state"), codes.state);
  }
  if (data.stability < 0) {
    addFieldIssue(ctx, fieldPath("stability"), codes.stability);
  }
  if (data.difficulty < DIFFICULTY_MIN || data.difficulty > DIFFICULTY_MAX) {
    addFieldIssue(ctx, fieldPath("difficulty"), codes.difficulty);
  }
  if (data.scheduledDays < 0) {
    addFieldIssue(ctx, fieldPath("scheduledDays"), codes.scheduledDays);
  }
  if (data.learningSteps < 0) {
    addFieldIssue(ctx, fieldPath("learningSteps"), codes.learningSteps);
  }
  if (codes.reps !== undefined && data.reps !== undefined && data.reps < 0) {
    addFieldIssue(ctx, fieldPath("reps"), codes.reps);
  }
  if (codes.lapses !== undefined && data.lapses !== undefined && data.lapses < 0) {
    addFieldIssue(ctx, fieldPath("lapses"), codes.lapses);
  }
}

export function validateReviewRating(rating: number, ctx: z.RefinementCtx, pathPrefix: (string | number)[] = []) {
  if (rating < RATING_MIN || rating > RATING_MAX) {
    addFieldIssue(ctx, [...pathPrefix, "rating"], "validation.reviews.rating");
  }
}

export function validateReviewTime(time: number, ctx: z.RefinementCtx, pathPrefix: (string | number)[] = []) {
  if (time < 0) {
    addFieldIssue(ctx, [...pathPrefix, "time"], "validation.reviews.time");
  }
}
